### PR TITLE
PersistentDict: concurrency improvements

### PR DESCRIPTION
**Pull request by matthiasdiener**
Should address the concurrency issues seen in #227.

I have a small concurrency test in https://github.com/matthiasdiener/skvlite/blob/7718f6729d0c6db8aabb047720f7c1f42f94689e/examples/concurrency.py which I could add here if you like @inducer.

_Please squash_

### CODE REVIEW

File: pytools/persistent_dict.py [hunk 1]
```diff
@@ -522,6 +522,25 @@ def _collision_check(self, key: K, stored_key: K) -> None:
...
+            except sqlite3.OperationalError as e:
+                # If the database is busy, retry
+                if (hasattr(e, "sqlite_errorcode")
+                        and not e.sqlite_errorcode == sqlite3.SQLITE_BUSY):
+                    raise
+                if n % 20 == 0:
+                    from warnings import warn
+                    warn(f"PersistentDict: database '{self.filename}' busy, {n} "
+                         "retries")
+            else:
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytools/persistent_dict.py:541]: We could add a small delay, but Imho I don't see much benefit - fairness between ranks seemed decent even without an extra delay, based on the results I mentioned yesterday:

image

[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytools/persistent_dict.py:541]: I'm not sure we can specify a particular timeout - any value will depend on a lot of other factors like system speed, FS contention, number of processes accessing the same dictionary ... We had issues with the previous behavior as well (see e.g. https://github.com/illinois-ceesd/mirgecom/pull/1042#issuecomment-2140028002), which is one of the reasons we made the caches rank-local.

**Comment by matthiasdiener** [comment]
#232 is the concurrency test that fails without this PR.

File: pytools/persistent_dict.py [hunk 2]
```diff
@@ -522,6 +522,25 @@ def _collision_check(self, key: K, stored_key: K) -> None:
             stored_key == key  # pylint:disable=pointless-statement  # noqa: B015
             raise NoSuchEntryCollisionError(key)
 
+    def _exec_sql(self, *args: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytools/persistent_dict.py]: Thanks! Added in a29e43d.

File: pytools/test/test_persistent_dict.py [hunk 3]
```diff
@@ -899,8 +899,109 @@ def method(self):
     # }}}
 
 
+# {{{ mpi test infrastructure
+
+def run_test_with_mpi(num_ranks, f, *args, extra_env_vars=None):
```
Comment (inducer) [inline pytools/test/test_persistent_dict.py:904]: MPI seems like a heavy-handed thing to use here. Wouldn't [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) suffice?

File: pytools/persistent_dict.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (inducer) [reply pytools/persistent_dict.py:544]: Fair enough.

[the remark this answers is not in the document]
Comment (inducer) [reply pytools/persistent_dict.py:544]: Fair, let's not.

File: pytools/test/test_persistent_dict.py (hunk 3 above)
Comment (matthiasdiener) [reply pytools/test/test_persistent_dict.py]: What do you think of https://github.com/inducer/pytools/pull/231/commits/cc9966cd155dd5e92d09063cb879ce617f331ffe? (I haven't worked with `multiprocessing` before, so I'm not sure if that code is idiomatic). The test still fails in #232 without the concurrency improvements here.

**Comment by inducer** [comment]
Thanks!

**Pull request merged**